IdentityCore/src/util/NSString+MSIDExtensions.m

```diff
@@ -95,26 +95,42 @@ - (NSString *)msidTrimmedString
     return [self stringByTrimmingCharactersInSet:set];
 }
 
+- (NSString *)msidURLDecode
+{
+    return CFBridgingRelease(CFURLCreateStringByReplacingPercentEscapes(kCFAllocatorDefault, (CFStringRef)self, CFSTR("")));
+}
 
 - (NSString *)msidWWWFormURLDecode
 {
     // Two step decode: first replace + with a space, then percent unescape
-    CFMutableStringRef decodedString = CFStringCreateMutableCopy( NULL, 0, (__bridge CFStringRef)self );
-    CFStringFindAndReplace( decodedString, CFSTR("+"), CFSTR(" "), CFRangeMake( 0, CFStringGetLength( decodedString ) ), kCFCompareCaseInsensitive );
+    NSString *replacedString = [self stringByReplacingOccurrencesOfString:@"+" withString:@" "];
+    return [replacedString msidURLDecode];
+}
+
+- (NSString *)msidURLEncode
+{
+    static NSCharacterSet* set = nil;
     
-    CFStringRef unescapedString = CFURLCreateStringByReplacingPercentEscapes( NULL,                    // Allocator
-                                                                                          decodedString,           // Original string
-                                                                                          CFSTR("")); // Encoding
-    CFRelease( decodedString );
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        
+        NSMutableCharacterSet *allowedSet = [[NSCharacterSet URLQueryAllowedCharacterSet] mutableCopy];
+        [allowedSet addCharactersInString:@" "];
+        [allowedSet removeCharactersInString:@"!$&'()*+,/:;=?@"];
+        
+        // This set is essentially english alphanumeric plus -_.~
+        set = allowedSet;
+    });
     
-    return CFBridgingRelease(unescapedString);
+    NSString *encodedString = [self stringByAddingPercentEncodingWithAllowedCharacters:set];
+    return [encodedString stringByReplacingOccurrencesOfString:@" " withString:@"%20"];
 }
 
-
 - (NSString *)msidWWWFormURLEncode
 {
+    // Same as percent encode, with exception of replacing space with + instead of %20
```

Comment:
nit: can we reuse same character set between msidWWWFormURLEncode and msidURLEncode?

Reply:
sure, made a helper method

Reply:
The character set can be created like this below. The current one works but its hard to know what is included in the urlQueryAllowedCharacterSet. Please see below:
dispatch_once(&onceToken, ^{
        
        NSMutableCharacterSet *allowedSet = [[NSCharacterSet alphanumericCharacterSet] mutableCopy];
        [allowedSet addCharactersInString:@"-._~ "];
        set = allowedSet;
    });